thunder/core/pytree.py

```diff
@@ -1,52 +1,101 @@
+    namespace=OPTREE_NAMESPACE,
 )
 
 
-def tree_flatten(args):
-    if type(args) not in {
-        dict,
-        list,
-        str,
-        int,
-        bool,
-        tuple,
-        torch.dtype,
-        float,
-        dtypes.floating,
-        dtypes.bool_,
-        devices.Device,
-        torch.memory_format,
-        type(None),
-        slice,
-        complex,
-        type,
-        type(Ellipsis),
-        torch.Size,
-    } and not isinstance(args, (ProxyInterface)):
+def tree_flatten(args, namespace=""):
+    if (
+        type(args)
+        not in {
+            dict,
+            list,
+            str,
+            int,
+            bool,
+            tuple,
+            torch.dtype,
+            float,
+            dtypes.floating,
+            dtypes.bool_,
+            devices.Device,
+            torch.memory_format,
+            type(None),
+            slice,
+            complex,
+            type,
+            type(Ellipsis),
+            torch.Size,
+        }
+        and not isinstance(args, (ProxyInterface))
+        and not dataclasses.is_dataclass(args)
+    ):
         raise TypeError(f"tree_flatten of type {type(args)} is not supported.")
-    return optree.tree_flatten(args, none_is_leaf=True)
+    return optree.tree_flatten(args, none_is_leaf=True, namespace=namespace)
 
 
-tree_map = partial(optree.tree_map, none_is_leaf=True)
+# This is required in the `torch_autograd` part of the code where we split forward and backward fn.
+# We want to be able to inspect `dataclass` containers to see if they contain proxy
+# while generating the split functions.
+tree_map = partial(optree.tree_map, none_is_leaf=True, namespace=OPTREE_NAMESPACE)
 
 
 def tree_unflatten(values, spec):
     return optree.tree_unflatten(spec, values)
 
 
-__all__ = ["tree_flatten", "tree_unflatten", "tree_map"]
+_registered_dataclasses = set()
+
+
+def register_pytree_node_dataclass(cls):
+    # We don't use `dataclasses.asdict` as it recursively flattens all data classes (including
+    # thunder internal ones like `VJPDual` (and also it is relatively slower as it calls copy.deepcopy()).
+    def unpack(cls) -> dict:
+        return {field.name: getattr(cls, field.name) for field in dataclasses.fields(cls)}
+
+    _flatten = lambda obj: tree_flatten(unpack(obj), namespace=OPTREE_NAMESPACE)
+    _unflatten = lambda spec, children: cls(**spec.unflatten(children))
+    optree.register_pytree_node(cls, _flatten, _unflatten, OPTREE_NAMESPACE)
+    return cls
+
+
+def _maybe_register_dataclass(t):
+    if dataclasses.is_dataclass(t) and t.__class__ not in _registered_dataclasses:
+        return True
+    return False
+
+
+# `tree_flatten_with_dataclass` iterates over the tree and registers functions to flatten dataclass objects present in the `tree`.
+# This is to facilitate peeking into the dataclass object to correctly get proxies when inspecting the BoundSymbols in the trace.
+def tree_flatten_with_dataclass(tree):
```

Comment:
Is there a situation where we'd prefer to use "tree_flatten_without_dataclasses"? Should "tree_flatten_with_dataclass" be just "tree_flatten"?

Reply:
IIRC, the `tree_flatten` usage in `codeutils.py` (used for printing/creating traces) requires `tree_flatten_without_dataclasses` as we want to handle a dataclass atomically (vs flattening it to a dict).

Ref: https://github.com/Lightning-AI/lightning-thunder/blob/a3e432f7174019b2eda85865890d5f7342a993c2/thunder/core/codeutils.py#L137-L143

Reply:
I think it would probably be better with dataclass construction symbol (trace would look cleaner and probably we won't need `tree_flatten_with_dataclass` as proxies will be arguments to this symbol). Will check this approach (and probably sync offline for more details). Thanks!